### PR TITLE
Fix psbt fuzz crash

### PR DESCRIPTION
**Pull request by sanket1729**
Fixes: https://github.com/rust-bitcoin/rust-bitcoin/issues/3628

This occurs when combining two PSBTs with different xpub key sources. Added a length check before indexing into slices to prevent out-of-bounds access.

For some reason, the precommit hook complained about non-ascii files. I don't think any of the names here are non-ascii

### CODE REVIEW

**Comment by apoelstra** [comment]
Heh, have the precommit hooks ever worked for you? When I try running the "ASCII check" even on master it finds a bunch of errant `0x0a`s, which are just line returns. And on your commit it finds many many more of them.

I suspect this depends on the details of your pager and other diff settings.

**Review by apoelstra — APPROVED** [review]
ACK 9aebb96fb99e8e9e019663659c6eff851a62f2ce; successfully ran local tests; thanks!

**Comment by tcharding** [comment]
I use the pre-commit hooks (well I almost always skip them with `-n` because I'm impatient).

In case its useful, in my gitconfig file I have

```

[core]
        pager = delta
        excludesfile = ~/.gitignore
        editor = /usr/bin/emacsclient

[merge]
	conflictstyle = diff3

[diff]
	colorMoved = zebra
        colorMovedWs = allow-indentation-change
```

And I have `$PAGER` set to `less -R`

**Review by tcharding — APPROVED** [review]
ACK 9aebb96fb99e8e9e019663659c6eff851a62f2ce

**Comment by apoelstra** [comment]
Interesting -- I also use `delta`, `diff3` and `less -R`. And `colorMoved`/`colorMovedWs` set to the same values. I'm not sure what's up.

One thing is that I have `delta --color-only` rather than just `delta`, but I don't think that should have any effect. Another difference is that I have `context = 25` which shows me many more context lines than the default 3.

**Comment by apoelstra** [comment]
You can set it by hand with `-U`. Like, `git diff -U3` gets the old behavior and `git diff -U10000` will show you entire files (usually :P) with the diffs highlighted.

**Comment by tcharding** [comment]
I use similar with grep all the time, never thought to look for it in diff - legend.

**Comment by sanket1729** [comment]
You have opened a whole new world for me. I was not even aware of the existence of pager and the fact that is configurable. `delta` is awesome

File: bitcoin/src/psbt/mod.rs [hunk 1]
```diff
@@ -2113,6 +2114,16 @@ mod tests {
         assert_eq!(psbt1, psbt2);
     }
 
+
+    // https://github.com/rust-bitcoin/rust-bitcoin/issues/3628
```
Comment (yancyribbens) [inline bitcoin/src/psbt/mod.rs:2118]: It's kind late since this is merged, but why is this issue link here?

Comment (sanket1729) [reply bitcoin/src/psbt/mod.rs:2118]: Just incase people want to reproduce the test files from the fuzz input.

Comment (sanket1729) [reply bitcoin/src/psbt/mod.rs:2118]: I could have been more explicit why I was linking that here.

Comment (apoelstra) [reply bitcoin/src/psbt/mod.rs:2118]: I think that a link to an issue immediately before a testcase is very clear. Though just having 3628 in the name was probably sufficient.

**Pull request merged**